packages/world/src/systems/MineSystem.sol

```diff
@@ -165,35 +176,44 @@ contract MineSystem is System {
           _removeBlock(relative, relativeCoord);
         }
       } else {
-        Mass._setMass(entityId, finalMass);
+        Mass._setMass(mined, finalMass);
       }
     }
 
     MineLib._requireMinesAllowed(caller, mineObjectTypeId, coord, extraData);
 
     if (finalMass == 0) {
       // Detach program if it exists
-      ProgramId program = entityId.getProgram();
+      ProgramId program = mined.getProgram();
       if (program.exists()) {
-        bytes memory onDetachProgram = abi.encodeCall(IDetachProgramHook.onDetachProgram, (caller, entityId, extraData));
+        bytes memory onDetachProgram = abi.encodeCall(IDetachProgramHook.onDetachProgram, (caller, mined, extraData));
         program.call({ gas: SAFE_PROGRAM_GAS, hook: onDetachProgram });
       }
 
       if (mineObjectTypeId == ObjectTypes.ForceField) {
-        destroyForceField(entityId);
+        destroyForceField(mined);
       }
     }
 
-    notify(caller, MineNotification({ mineEntityId: entityId, mineCoord: coord, mineObjectTypeId: mineObjectTypeId }));
+    notify(caller, MineNotification({ mineEntityId: mined, mineCoord: coord, mineObjectTypeId: mineObjectTypeId }));
 
-    return entityId;
+    return mined;
+  }
+
+  function mineUntilDestroyed(EntityId caller, Vec3 coord, uint16 toolSlot, bytes calldata extraData) public payable {
+    uint128 massLeft = 0;
+    do {
+      // TODO: factor out the mass reduction logic so it's cheaper to call
+      EntityId entityId = _mine(caller, coord, toolSlot, extraData);
+      massLeft = Mass._getMass(entityId);
+    } while (massLeft > 0);
   }
 
   function mineUntilDestroyed(EntityId caller, Vec3 coord, bytes calldata extraData) public payable {
```

Comment:
We should do the same for `mine` and `mineUntilDestroyed` for tools and not tools. Either use overloaded functions or add a suffix (vs doing both) for consistency

Reply:
done using overloaded functions, let me know if I should just use different names

Reply:
can keep this for now and do different names if its causing issues